libraries/AP_SmartAudio/AP_SmartAudio.cpp

```diff
@@ -0,0 +1,947 @@
+#include "AP_SmartAudio.h"
+#include <AP_Math/crc.h>
+#include <GCS_MAVLink/GCS.h>
+
+#define AP_VTX_DEV_MODE     // DEV TESTS ACTIVATE
+#ifdef AP_VTX_DEV_MODE
+# define debug(fmt, args...)	hal.console->printf("\r " fmt "\n", ##args)
+//#define debug(fmt,args...) gcs().send_text(MAV_SEVERITY_DEBUG , fmt, ##args);
+#else
+//# define debug(fmt, args...)	do {} while(0)
+#endif
+
+extern const AP_HAL::HAL &hal;
+const char *TAG="VTX-SMARTAUDIO ";
+
+
+// table of user settable parameters
+const AP_Param::GroupInfo AP_SmartAudio::var_info[] = {
+
+    // @Param: ENABLED
+    // @DisplayName: Enable SmartAudio protocol on VTX
+    // @Description: Enable SmartAudio protocol on VTX
+    // @Values: 0:Disabled,1:Enabled
+    // @User: Advanced
+    AP_GROUPINFO_FLAGS("ENABLED",     1, AP_SmartAudio::configured_smartaudio_params, enabled, 0,AP_PARAM_FLAG_ENABLE),
+
+    // @Param: VERSION
+    // @DisplayName: SmartAudio version
+    // @Description: SmartAudio version
+    // @Values: 0: Version 1,1:Version 2, 3: Version 2.1
+    // @User: Advanced
+    AP_GROUPINFO("VERSION",     2, AP_SmartAudio::configured_smartaudio_params, protocol_version, 0),
+
+    // @Param: DEFAULTS
+    // @DisplayName: VTX will be configured with VTX params defined by default.
+    // @Description: VTX will be configured with VTX params defined by default.
+    // @Values: 0: Disabled,1: Enabled
+    // @User: Advanced
+    AP_GROUPINFO("DEFAULTS",     3, AP_SmartAudio::configured_smartaudio_params, setup_defaults, 0),
+
+    AP_GROUPEND
+
+};
+/**
+ * Constructor
+ */
+AP_SmartAudio::AP_SmartAudio()
+{
+    AP_Param::setup_object_defaults(this, var_info);
+    singleton = this;
+
+}
+
+AP_SmartAudio *AP_SmartAudio::singleton;
+AP_VideoTX *ap_video_tx=AP_VideoTX::get_singleton();
+
+// initialization start making a request settings to the vtx
+bool AP_SmartAudio::init()
+{
+        
+        debug("SmartAudio init");
+
+    if(smartaudio_params.enabled.get()==0){
+        
+        debug("SmartAudio protocol it's not active");
+        return false;
+    }
+        
+        debug("SmartAudio protocol is active");
+
+    // init uart
+    _port = AP::serialmanager().find_serial(AP_SerialManager::SerialProtocol_SmartAudio, 0);
+    if(_port!=nullptr){
+
+          if (!hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&AP_SmartAudio::loop, void),
+                                          "SmartAudio",
+                                          1024, AP_HAL::Scheduler::PRIORITY_IO, 60)) {
+            return false;
+            }
+
+        // setup port options
+        _port->set_stop_bits(2);
+        _port->set_flow_control(AP_HAL::UARTDriver::FLOW_CONTROL_DISABLE);
+        _port->set_options(AP_HAL::UARTDriver::OPTION_HDPLEX);
```

Comment:
Usually this is left to the user to configure on SERIALx_OPTIONS. Maybe this is ok. If you are going to do this you probably want to set the baud rate as well. I think you also probably don't want blocking writes

Reply:
As i understand, it's seems a design aspect of the spec, so it's hardcoded to talk correctly with the uart. Anyway if you think the other aproach give us some value. I'll remove this.

Reply:
I'm really not sure. I guess what happens if someone swaps the TX/RX?

Reply:
Good point, i'll research this way.

Reply:
uart params could be mixed? Example: HALFDUPLEX AND SWAPPING ?